lib/Sema/CSDiag.cpp

```diff
@@ -3659,10 +3684,33 @@ bool FailureDiagnosis::visitApplyExpr(ApplyExpr *callExpr) {
     // Re-type-check the argument with the expected type of the candidate set.
     // This should produce a specific and tailored diagnostic saying that the
     // type mismatches with expectations.
-    if (!typeCheckChildIndependently(badArgExpr,
-                                     calleeInfo.failedArgument.parameterType,
+    Type paramType = calleeInfo.failedArgument.parameterType;
+    if (!typeCheckChildIndependently(badArgExpr, paramType,
                                      CTP_CallArgument, options))
       return true;
+      
+    // If that fails, it could be that the argument doesn't conform to an archetype.
+    Type argType = badArgExpr->getType();
+
+    if (paramType->is<ArchetypeType>() && !argType->hasTypeVariable() &&
```

Comment:
Did the cleanup suggestions. Don't think this should be moved into diagnoseSimpleErrors() because the call site at visitUnresolvedMemberExpr() has explicitly narrowed down causes to the existing checks in simple errors already. 

But it is cleaner as a member on CalleeCandidateInfo, like diagnoseSimpleErrors. Did that. I need to grok visitSubscriptExpr() before adding a call site there.